benchmarks/src/jmh/java/io/opencensus/benchmarks/trace/SpanOperationsBenchmark.java

```diff
@@ -0,0 +1,221 @@
+                SpanContext.create(
+                    TraceId.fromBytes(
+                        new byte[] {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, (byte) i}),
+                    SpanId.fromBytes(new byte[] {1, 2, 3, 4, 5, 6, 7, (byte) i}),
+                    TraceOptions.DEFAULT,
+                    TRACESTATE_DEFAULT),
+                Link.Type.PARENT_LINKED_SPAN);
+      }
+    }
+  }
+
+  /** Add attributes individually. */
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public Span putAttribute(Data data) {
+    Span span = data.attributeSpan;
+    for (int i = 0; i < data.size; i++) {
+      span.putAttribute(data.attributeKeys[i], data.attributeValues[i]);
+    }
+    return span;
+  }
+
+  /** Add attributes as a map. */
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public Span putAttributes(Data data) {
+    Span span = data.attributeSpan;
+    span.putAttributes(data.attributeMap);
+    return span;
+  }
+
+  /** Add an annotation as description only. */
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public Span addAnnotationEmpty(Data data) {
+    Span span = data.annotationSpanEmpty;
+    span.addAnnotation(ANNOTATION_DESCRIPTION);
+    return span;
+  }
+
+  /** Add an annotation with attributes. */
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public Span addAnnotationWithAttributes(Data data) {
+    Span span = data.annotationSpanAttributes;
+    span.addAnnotation(ANNOTATION_DESCRIPTION, data.attributeMap);
+    return span;
+  }
+
+  /** Add an annotation with an annotation. */
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public Span addAnnotationWithAnnotation(Data data) {
+    Span span = data.annotationSpanAnnotation;
+    Annotation annotation =
+        Annotation.fromDescriptionAndAttributes(ANNOTATION_DESCRIPTION, data.attributeMap);
+    span.addAnnotation(annotation);
+    return span;
+  }
+
+  /** Add message events. */
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public Span addMessageEvent(Data data) {
+    Span span = data.messageEventSpan;
+    for (int i = 0; i < data.size; i++) {
+      span.addMessageEvent(data.messageEvents[i]);
+    }
+    return span;
+  }
+
+  /** Add links. */
+  @Benchmark
+  @BenchmarkMode(Mode.AverageTime)
+  @OutputTimeUnit(TimeUnit.NANOSECONDS)
+  public Span addLink(Data data) {
+    Span span = data.linkSpan;
+    for (int i = 0; i < data.size; i++) {
+      span.addLink(data.links[i]);
+    }
+    return span;
+  }
+
+  private static AttributeValue[] createAttributeValues(int size) {
+    AttributeValue[] attributeValues = new AttributeValue[size];
+    for (int i = 0; i < size; i++) {
+      attributeValues[i] = AttributeValue.stringAttributeValue(ATTRIBUTE_VALUE + "-i");
```

Comment:
same as before. should it be ATTRIBUTE_VALUE + "-" + i ?

Reply:
Good catch - that's actually in the trace benchmarks, which seem to have appeared here with the rebase merge commit issue. I'll put together a new pr to fix the issue.